GSoC contributor blog for Third-party repository detection probe (#6538)
jenkins-infra/jenkins.io: content/blog/2023/07/16/2023-07-16-third-party-repository-detection-probe.adoc

---
layout: post
title: Third Party Repository Detection Probe
tags:
- gsoc
- gsoc2023
- healthscore
- probes
- plugin
authors:
- jagruti
opengraph:
  image: /images/gsoc/2023/jagruti/blog_20230716.png
sig:
- gsoc
---

== Introduction

> A third-party repository is a repository that is not hosted at https://repo.jenkins-ci.org/.
> 
> The Jenkins Infra team was concerned about the dependencies used from the third-party repositories. These repositories are concern not just for security reasons but also for reliability.
> 
> Third party repository detection probe ensures that plugins are built only from trusted and reliable Maven repositories.

Reviewer: ```suggestion
A third-party repository is a repository that is not hosted at https://repo.jenkins-ci.org/.

The Jenkins Infra team was concerned about the dependencies used from the third-party repositories.
These repositories are a concern, not just for security reasons, but also for reliability.

The third-party repository detection probe ensures that plugins are built only from trusted and reliable Maven repositories.
```

> == Importance of the probe
> This is a community requested probe. In the ticket description for this probe, one of the plugin developers had asked the Jenkins team to include a third-party repository in the plugin build. With the probe in place the Jenkins team would already know the plugins they should add to the build.

Reviewer: ```suggestion
== Importance of the probe

This probe was requested by the community.
In the ticket description, one of the plugin developers asked the Jenkins team to include a third-party repository in the plugin build.
With the probe in place, the Jenkins team can now identify the plugins they should add to the build process.
```

> == Challenges
> In a Maven project all the dependencies are listed the `pom.xml` file. A plugin can have multiple pom files. Each module may have its own pom. Which means they can be multiple child poms in the same repository.

Reviewer: ```suggestion
== Challenges

In a Maven project, all the dependencies are listed in the `pom.xml` file.
A plugin can have multiple pom files, and each module may have its own pom.
This means that there can be multiple child poms in the same repository.
```

> A major challenge in this probe was to consider parent and child pom relationship from all angles.

Reviewer: ```suggestion
A major challenge in this probe was to consider the parent and child pom relationship from all angles.
```

> == Outcome
> During code reviews and test cases, we noticed there were assumptions made regarding Maven's project hierarchies.
> 
> There were edge case scenarios that were not considered while specifying the probe. Considering the number of open questions, this probe is put on hold until further study was made to adapt the project architecture to tackle the new identified cases.

Reviewer: ```suggestion
== Outcome

During code reviews and test cases, we noticed that there were assumptions made regarding Maven's project hierarchies.

Furthermore, some edge-case scenarios were not considered while specifying the probe.
Due to the number of open questions, this probe was put on hold until further study was made to adapt the project architecture to tackle the newly identified cases.
```

> == Learning
> Implemented parameterized test cases to test the probe. Read the Maven documentation to test POM structure, inheritance, and Maven API.

Reviewer: ```suggestion
== Learning

I implemented parameterized test cases to test the probe.
Additionally, I extensively read the Maven documentation to understand how to test POM structure, inheritance, and Maven API.
```

> == Conclusion
> This is a partially completed probe. 60% of the code and test cases is completed.
> 
> The idea is to develop an effective-pom-probe that will compile all the parent-child pom relationships in one file. We can then list third-party-repositories from every pom file used in the repository.

Reviewer: ```suggestion
== Conclusion
This probe is currently partially completed, with approximately 60% of the code and test cases already implemented.

The objective is to create an efficient pom probe capable of aggregating all the parent-child pom relationships into a single file.
By doing so, we can easily list the third-party repositories used in every pom file within the repository.
```

> == Links
> For more information or to find answers to any questions you might have, head to the official GSoC 2023 project link:/projects/gsoc/2023/projects/add-probes-to-plugin-health-score/[Adding Probes to  "Plugin Health Score"] description page.

Reviewer: ```suggestion
== Links

For more information or to find answers to any questions you might have, please visit the official GSoC 2023 project link:/projects/gsoc/2023/projects/add-probes-to-plugin-health-score/[Adding Probes to "Plugin Health Score"] description page.
```

---
layout: post
title: Third Party Repository Detection Probe
tags:
- gsoc
- gsoc2023
- healthscore
- probes
- plugin
authors:
- jagruti
opengraph:
  image: /images/gsoc/2023/jagruti/blog_20230716.png
sig:
- gsoc
---

== Introduction

> A third-party repository is a repository that is not hosted at https://repo.jenkins-ci.org/.

Reviewer 2: ```suggestion
A third-party repository is a repository that is not hosted at link:https://repo.jenkins-ci.org/[].
```
adding link macro

> The Jenkins Infra team was concerned about the dependencies used from the third-party repositories.

Reviewer 2: ```suggestion
The Jenkins Infra team was concerned about the dependencies used by third-party repositories.
```
suggestion for wording to read more clearly

These repositories are a concern, not just for security reasons, but also for reliability.

> The third-party repository detection probe ensures that plugins are built only from trusted and reliable Maven repositories.

Reviewer 2: ```suggestion
The third-party repository detection probe ensures that plugins are built _only_ from trusted and reliable Maven repositories.
```
suggestion for formatting/italics

== Importance of the probe

This probe was requested by the community.
In the ticket description, one of the plugin developers asked the Jenkins team to include a third-party repository in the plugin build.
With the probe in place, the Jenkins team can now identify the plugins they should add to the build process.

== Challenges

In a Maven project, all the dependencies are listed in the `pom.xml` file.
A plugin can have multiple pom files, and each module may have its own pom.
This means that there can be multiple child poms in the same repository.

> A major challenge in this probe was to consider the parent and child pom relationship from all angles.

Reviewer 2: ```suggestion
A major challenge in this probe was considering the parent and child pom relationship from all angles.
```
suggestion for wording/action

== Outcome

During code reviews and test cases, we noticed that there were assumptions made regarding Maven's project hierarchies.

Furthermore, some edge-case scenarios were not considered while specifying the probe.

> Due to the number of open questions, this probe was put on hold until further study was made to adapt the project architecture to tackle the newly identified cases.

Reviewer 2: ```suggestion
Due to the number of open questions, this probe was put on hold until further research was completed.
The goal of doing this was to adapt the project architecture to tackle the newly identified cases.
```
suggestion to add some text here to flow a bit better
